src/client/websocket/WebSocketShard.js

```diff
@@ -0,0 +1,362 @@
+      this.debug('Received null packet');
+      return false;
+    }
+
+    switch (packet.t) {
+      case WSEvents.READY:
+        this.sessionID = packet.d.session_id;
+        this.trace = packet.d._trace;
+        this.status = Status.READY;
+        this.debug(`READY ${this.trace.join(' -> ')} ${this.sessionID}`);
+        this.heartbeat();
+        break;
+      case WSEvents.RESUMED: {
+        this.trace = packet.d._trace;
+        this.status = Status.READY;
+        const replayed = packet.s - this.sequence;
+        this.debug(`RESUMED ${this.trace.join(' -> ')} | replayed ${replayed} events.`);
+        this.heartbeat();
+        break;
+      }
+    }
+
+    if (packet.s > this.sequence) this.sequence = packet.s;
+
+    switch (packet.op) {
+      case OPCodes.HELLO:
+        this.identify();
+        return this.heartbeat(packet.d.heartbeat_interval);
+      case OPCodes.RECONNECT:
+        return this.reconnect();
+      case OPCodes.INVALID_SESSION:
+        if (!packet.d) this.sessionID = null;
+        this.sequence = -1;
+        this.debug('Session invalidated');
+        return this.reconnect();
+      case OPCodes.HEARTBEAT_ACK:
+        return this.ackHeartbeat();
+      case OPCodes.HEARTBEAT:
+        return this.heartbeat();
+      default:
+        return this.manager.handlePacket(packet, this);
+    }
+  }
+
+  /**
+   * Called whenever a connection is opened to the gateway.
+   */
+  onOpen() {
+    this.debug('Connection open');
+  }
+
+  /**
+   * Called whenever a message is received.
+   * @param {Event} event Event received
+   */
+  onMessage({ data }) {
+    if (data instanceof ArrayBuffer) data = new Uint8Array(data);
+    const l = data.length;
+    const flush = l >= 4 &&
+      data[l - 4] === 0x00 &&
+      data[l - 3] === 0x00 &&
+      data[l - 2] === 0xFF &&
+      data[l - 1] === 0xFF;
+
+    this.inflate.push(data, flush && zlib.Z_SYNC_FLUSH);
+    if (!flush) return;
+    let packet;
+    try {
+      packet = WebSocket.unpack(this.inflate.result);
+    } catch (err) {
+      this.manager.client.emit(Events.ERROR, err);
+      return;
+    }
+    if (packet.t === 'READY') {
+      this.emit('ready');
+    }
+    this.onPacket(packet);
+  }
+
+  /**
+   * Called whenever an error occurs with the WebSocket.
+   * @param {Error} error The error that occurred
+   */
+  onError(error) {
+    if (error && error.message === 'uWs client connection error') {
+      this.reconnect();
+      return;
+    }
+    this.manager.client.emit(Events.ERROR, error);
+  }
+
+  /**
+   * @external CloseEvent
+   * @see {@link https://developer.mozilla.org/en-US/docs/Web/API/CloseEvent}
+   */
+
+  /**
+   * Called whenever a connection to the gateway is closed.
+   * @param {CloseEvent} event Close event that was received
+   */
+  onClose(event) {
+    this.emit('close', event);
+    if (event.code === 1000 ? this.expectingClose : WSCodes[event.code]) {
+      this.manager.client.emit(Events.DISCONNECT, event);
```

Comment:
Shouldn't this also inform the user what shard disconnected? (like `resumed` does)